### ISSUE TITLE
anvil `TypedTransaction` fastrlp decoding of `Create` transactions fails

### ISSUE THREAD

**Issue by sveitser** [foundry-rs/foundry#2965]
### Component

Anvil

### Have you ensured that all of these are up to date?

- [ ] Foundry
- [ ] Foundryup

### What version of Foundry are you on?

4e11d1ff1d01cec67399cb7fdaaf7fa3cf467532

### Operating System

Linux

### Describe the bug

I have tried to switch over from using `rlp` to `fastrlp` but I get failures every time I decode a `create` transactions.

Below is an example.

A Create tx 
- https://goerli.etherscan.io/tx/0x7d82c63a0d24b4622cb8f836fff374f64578693b2dfc3db5f53963ecd063674f
- https://goerli.etherscan.io/getRawTx?tx=0x7d82c63a0d24b4622cb8f836fff374f64578693b2dfc3db5f53963ecd063674f

```rust
let tx_bytes =
      hex::decode("02f901ee05228459682f008459682f11830209bf8080b90195608060405234801561001057600080fd5b50610175806100206000396000f3fe608060405234801561001057600080fd5b506004361061002b5760003560e01c80630c49c36c14610030575b600080fd5b61003861004e565b604051610045919061011d565b60405180910390f35b60606020600052600f6020527f68656c6c6f2073746174656d696e64000000000000000000000000000000000060405260406000f35b600081519050919050565b600082825260208201905092915050565b60005b838110156100be5780820151818401526020810190506100a3565b838111156100cd576000848401525b50505050565b6000601f19601f8301169050919050565b60006100ef82610084565b6100f9818561008f565b93506101098185602086016100a0565b610112816100d3565b840191505092915050565b6000602082019050818103600083015261013781846100e4565b90509291505056fea264697066735822122051449585839a4ea5ac23cae4552ef8a96b64ff59d0668f76bfac3796b2bdbb3664736f6c63430008090033c080a0136ebffaa8fc8b9fda9124de9ccb0b1f64e90fbd44251b4c4ac2501e60b104f9a07eb2999eec6d185ef57e91ed099afb0a926c5b536f0155dd67e537c7476e1471")
          .unwrap();
anvil_core::eth::transaction::TypedTransaction::decode(&mut &tx_bytes[..]).unwrap();
```

Fails with
```
thread 'test::test_decode' panicked at 'called `Result::unwrap()` on an `Err` value: UnexpectedString
```

For other `create` transactions the `DecodeError` returned varies. It also affects Legacy as well as `0x01` and `0x02` transactions.

**Comment by mattsse** [comment]
`rlp::decode` works?

**Comment by sveitser** [comment]
Yes that works.

**Comment by sveitser** [comment]
No sorry, it also fails in this example with `ExpectedToBeList`. Is this expected because it's not a valid RLP encoding because of the version prefix?

**Comment by sveitser** [comment]
I will try come up with a better example.

**Comment by mattsse** [comment]
ref 

https://github.com/foundry-rs/foundry/blob/863561af4bec68dcd62a99672574be4577ea3855/anvil/src/eth/api.rs#L736-L745

**Comment by sveitser** [comment]
Thanks. I should not have picked that example. I'm still not quite sure if there isn't an issue with the `fastrlp` decoding.

Below are a toy `Call` and `Create` tx encoded and then decoded once each with `rlp` and `fastrlp`. The `fastrlp` decode step fails for the `Create` transaction, the other 3 tests pass.

```rust
#[cfg(test)]
mod test2 {
    use anvil_core::eth::transaction::{EIP2930Transaction, TransactionKind, TypedTransaction};
    use ethereum_types::{Address, H256, U256};
    use ethers_core::types::Bytes;

    // passes
    #[test]
    fn test_decode_rlp_create() {
        let tx = TypedTransaction::EIP2930(EIP2930Transaction {
            chain_id: 1u64,
            nonce: U256::from(0),
            gas_price: U256::from(1),
            gas_limit: U256::from(2),
            kind: TransactionKind::Create,
            value: U256::from(3),
            input: Bytes::from(vec![1, 2]),
            odd_y_parity: true,
            r: H256::default(),
            s: H256::default(),
            access_list: vec![].into(),
        });

        let encoded = rlp::encode(&tx);
        let decoded: TypedTransaction = rlp::decode(&encoded).unwrap();
        assert_eq!(decoded, tx);
    }

    // passes
    #[test]
    fn test_decode_rlp_call() {
        let tx = TypedTransaction::EIP2930(EIP2930Transaction {
            chain_id: 1u64,
            nonce: U256::from(0),
            gas_price: U256::from(1),
            gas_limit: U256::from(2),
            kind: TransactionKind::Call(Address::default()),
            value: U256::from(3),
            input: Bytes::from(vec![1, 2]),
            odd_y_parity: true,
            r: H256::default(),
            s: H256::default(),
            access_list: vec![].into(),
        });

        let encoded = rlp::encode(&tx);
        let decoded: TypedTransaction = rlp::decode(&encoded).unwrap();
        assert_eq!(decoded, tx);
    }

    // fails
    #[test]
    fn test_decode_fastrlp_create() {
        use bytes::BytesMut;
        use fastrlp::{Decodable, Encodable};

        let tx = TypedTransaction::EIP2930(EIP2930Transaction {
            chain_id: 1u64,
            nonce: U256::from(0),
            gas_price: U256::from(1),
            gas_limit: U256::from(2),
            kind: TransactionKind::Create,
            value: U256::from(3),
            input: Bytes::from(vec![1, 2]),
            odd_y_parity: true,
            r: H256::default(),
            s: H256::default(),
            access_list: vec![].into(),
        });

        let mut encoded = BytesMut::new();
        tx.encode(&mut encoded);

        let decoded = TypedTransaction::decode(&mut &*encoded).unwrap();
        // XXX fails with: `Err` value: UnexpectedString',
        assert_eq!(decoded, tx);
    }

    // passes
    #[test]
    fn test_decode_fastrlp_call() {
        use bytes::BytesMut;
        use fastrlp::{Decodable, Encodable};

        let tx = TypedTransaction::EIP2930(EIP2930Transaction {
            chain_id: 1u64,
            nonce: U256::from(0),
            gas_price: U256::from(1),
            gas_limit: U256::from(2),
            kind: TransactionKind::Call(Address::default()),
            value: U256::from(3),
            input: Bytes::from(vec![1, 2]),
            odd_y_parity: true,
            r: H256::default(),
            s: H256::default(),
            access_list: vec![].into(),
        });

        let mut encoded = BytesMut::new();
        tx.encode(&mut encoded);

        let decoded = TypedTransaction::decode(&mut &*encoded).unwrap();
        assert_eq!(decoded, tx);
    }
}
```

**Comment by mattsse** [comment]
perhaps the issue is here

https://github.com/foundry-rs/foundry/blob/863561af4bec68dcd62a99672574be4577ea3855/anvil/core/src/eth/transaction/mod.rs#L178-L187

**Comment by sveitser** [comment]
But I think the `rlp` code is working fine.

In the following code the `buf` is passed in as mutable but `.first()` does not seem to mutate the buffer. Could this be the problem?

https://github.com/foundry-rs/foundry/blob/863561af4bec68dcd62a99672574be4577ea3855/anvil/core/src/eth/transaction/mod.rs#L205-L218

In the following snippet `encoded` is still the same (b"\x80") at the end:

```rust
use bytes::BytesMut;
use fastrlp::{Decodable, Encodable};
let kind = TransactionKind::Create;
let mut encoded = BytesMut::new();
kind.encode(&mut encoded);
println!("encoded: {:?}", encoded);

let decoded = TransactionKind::decode(&mut &*encoded).unwrap();
assert_eq!(kind, decoded);
println!("encoded after: {:?}", encoded);
```

**Comment by sveitser** [comment]
My previous snippet does not consume what's in `encoded` anyway but I think the one below might illustrate a problem that the empty string of `Create` is not consumed.

```rust
use bytes::BytesMut;
use fastrlp::{Decodable, Encodable};
let kind = TransactionKind::Create;
let mut encoded = BytesMut::new();
kind.encode(&mut encoded);
let encoded: &mut &[u8] = &mut &*encoded;
println!("create encoded before: {:?}", encoded);

let decoded = TransactionKind::decode(encoded).unwrap();
assert_eq!(kind, decoded);
println!("create encoded after : {:?}", encoded);

let kind = TransactionKind::Call(Address::default());
let mut encoded = BytesMut::new();
kind.encode(&mut encoded);
let encoded: &mut &[u8] = &mut &*encoded;
println!("call encoded before: {:?}", encoded);

let decoded = TransactionKind::decode(encoded).unwrap();
assert_eq!(kind, decoded);
println!("call encoded after : {:?}", encoded);
```

Prints
```
create encoded before: [128]
create encoded after : [128]
call encoded before: [148, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
call encoded after : []
```

**Comment by mattsse** [comment]
cc @Rjected

**Comment by Rjected** [comment]
Yes, the problem is that the 0x80 byte is not being consumed. good catch, here's a fix: #2968!
